numba/tests/npyufunc/test_gufunc.py

```diff
@@ -68,6 +68,78 @@ class TestGUFuncParallel(TestGUFunc):
     target = 'parallel'
 
 
+class TestDynamicGUFunc(TestCase):
+    target = 'cpu'
+
+    def test_dynamic_matmul(self):
+
+        def check_matmul_gufunc(gufunc, A, B, C):
+            gufunc(A, B, C)
+            Gold = ut.matrix_multiply(A, B)
+            np.testing.assert_allclose(C, Gold, rtol=1e-5, atol=1e-8)
+
+        gufunc = GUVectorize(matmulcore, '(m,n),(n,p)->(m,p)',
+                             target=self.target)
+
+        matrix_ct = 10
+        Ai64 = np.arange(matrix_ct * 2 * 4, dtype=np.int64).reshape(matrix_ct, 2, 4)
+        Bi64 = np.arange(matrix_ct * 4 * 5, dtype=np.int64).reshape(matrix_ct, 4, 5)
+        Ci64 = np.arange(matrix_ct * 2 * 5, dtype=np.int64).reshape(matrix_ct, 2, 5)
+        check_matmul_gufunc(gufunc, Ai64, Bi64, Ci64)
+
+        A = np.arange(matrix_ct * 2 * 4, dtype=np.float32).reshape(matrix_ct, 2, 4)
+        B = np.arange(matrix_ct * 4 * 5, dtype=np.float32).reshape(matrix_ct, 4, 5)
+        C = np.arange(matrix_ct * 2 * 5, dtype=np.float32).reshape(matrix_ct, 2, 5)
+        check_matmul_gufunc(gufunc, A, B, C)  # trigger compilation
+
+        self.assertEqual(len(gufunc.types), 2)  # ensure two versions of gufunc
+
+
+    def test_dynamic_ufunc_like(self):
+
+        def check_ufunc_output(gufunc, x):
+            out = np.zeros(10, dtype=x.dtype)
+            gufunc(x, x, x, out)
+            np.testing.assert_equal(out, x * x + x)
+
+        # Test problem that the stride of "scalar" gufunc argument not properly
+        # handled when the actual argument is an array,
+        # causing the same value (first value) being repeated.
+        gufunc = GUVectorize(axpy, '(), (), () -> ()', target=self.target)
+        x = np.arange(10, dtype=np.intp)
+        check_ufunc_output(gufunc, x)
+
+
+    def test_dynamic_scalar_output(self):
+        """
+        Note that scalar output is a 0-dimension array that acts as
+        a pointer to the output location.
+        """
+
+        @guvectorize('(n)->()', target=self.target, nopython=True)
+        def sum_row(inp, out):
+            tmp = 0.
+            for i in range(inp.shape[0]):
+                tmp += inp[i]
+            out[()] = tmp
+
```

Comment:
one would expect the following:
```suggestion
        self.assertTrue(sum_row.is_dynamic)
```
but it is failing.

Reply:
Hi @sklam, I've improved `is_dynamic` behavior in my last commit. One thing that I notice is that some tests uses `GUVectorize` class directly instead of the `@guvectorize` decorator:

https://github.com/numba/numba/blob/ed4dac1ff741341c14b6e75ad1d97b981d999301/numba/tests/npyufunc/test_gufunc.py#L38-L43

https://github.com/numba/numba/blob/ed4dac1ff741341c14b6e75ad1d97b981d999301/numba/tests/npyufunc/test_gufunc.py#L81-L88

And in those cases it is harder to guess if types will be added later via `.add()` method or if the gufunc will be dynamic.